### DISCUSSION TITLE
best_FL_global_model.pt is selected from n-1 rounds of learning

### DISCUSSION THREAD

**Discussion by parkeraddison** [NVIDIA/NVFlare#802] [Q&A] [answered]
It looks like the best_FL_global_model.pt is only ever selected/saved in a round prior to aggregation (EventType.BEFORE_AGGREGATE), using the initial metrics (MetaKey.INITIAL_METRICS) computed before training took place at the client sites (ValidateType.BEFORE_TRAIN_VALIDATE).

This essentially means that the selection of a "best" FL global model can only take place on the n-1th round's global model, so if the global model improved in the nth round of learning then during cross-site eval we see that SRV_FL_global_model.pt actually outperforms SRV_best_FL_global_model.pt.

I'm wondering, could there be an additional event fired or method called after the rounds are complete to perform one final validation of the current FL global model across all sites and fire EventType.GLOBAL_BEST_MODEL_AVAILABLE accordingly?

**Comment by holgerroth** [comment, marked as answer]
Yes, your observation is correct. You could add a task to [workflow](https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_common/workflows/scatter_and_gather.py) that does a final evaluation before completing the run.

**Comment by pygabc1** [comment]
Hi, does it mean that the metric (i.e., accuracy) of SRV_best_FL_global_model could be worse than the metric of SRV_FL_global_model?  I have this situation now.

**Comment by holgerroth** [comment]
It could be if your model is still converging. Typically the best global model selection is useful to avoid overfitting the training data, assuming the validation performance will decrease later in training.

**Reply by pygabc1** [reply]
Hi, thank you for previous reply, and I have one more question. 
Printing the following lines in cifar10_learner.py (in fedavg, fedopt, fedprox in cifar10) shows saving the only model with better accuracy as the epoch increase, but SRV_FL_global_model.pt or SRV_best_FL_global_model.pt after maximum epoch is NOT the best model.  For example, assume that the best model is gotten in epoch 10 and the maximum epoch is 20, and then, the final SRV_FL_global_model.pt or SRV_best_FL_global_model.pt should be the best obtained from epoch 10, but the final model is from epoch 20. Why? I tried multiple times. It is linux OS and 3.8 python, and 2.2 NVFlare version.
>             if val_freq > 0 and epoch % val_freq == 0:
>                 acc = self.local_valid(self.valid_loader, abort_signal, tb_id="val_acc_local_model", fl_ctx=fl_ctx)
>                 if acc > self.best_acc:
>                     self.best_acc = acc
>                     self.save_model(is_best=True)

**Reply by holgerroth** [reply]
Not sure if I understand the issue correctly. The global model is always the one after aggregation. It's not any of the locally best models. So, `SRV_FL_global_model.pt` is the final aggregated one, and `SRV_best_FL_global_model.pt` is the aggregated one from the round where the average validation scores from the clients were the highest based on the `ModelSelector` criterion. This might be the n-1 round or an earlier round.

**Reply by holgerroth** [reply]
FYI, I just filed an issue for the CIFAR-10 Learner but I don't think it is related to your question https://github.com/NVIDIA/NVFlare/issues/1714

**Reply by pygabc1** [reply]
I am explaining more to help you, assuming the maximum epoch of 20 and central=True.  Due to above issue, I do two steps. First, I obtain "the best epoch (say, 10)" by printing the following line (which is saving only the better model for each epoch. In other words, NOT saving worse model than previous epoch). Second, I re-run again with maximum epoch 10 to get the best model. Thanks a lot.
```
    def local_train(self, fl_ctx, train_loader, model_global, abort_signal: Signal, val_freq: int = 0):
        for epoch in range(self.aggregation_epochs):
......
            if val_freq > 0 and epoch % val_freq == 0:
                acc = self.local_valid(self.valid_loader, abort_signal, tb_id="val_acc_local_model", fl_ctx=fl_ctx)
                if acc > self.best_acc:
                    self.best_acc = acc
                    self.save_model(is_best=True)
```

**Reply by holgerroth** [reply]
You shouldn't need to run twice. The best local models are saved on the client workspace as "best_local_model.pt".

**Reply by pygabc1** [reply]
> You shouldn't need to run twice. The best local models are saved on the client workspace as "best_local_model.pt".

Thank you for reply. Well, the issue is that the test performance of trained-model saved by 20 epoch by test data is much worse than the test performance of the one saved by 10 epoch (optimal epoch).

**Reply by holgerroth** [reply]
You mean the global model at epoch 20, correct? For local models, you should get both the best and the final one as well in the client working directories.

**Reply by pygabc1** [reply]
> You mean the global model at epoch 20, correct? For local models, you should get both the best and the final one as well in the client working directories.

Yes, the test performance by the global-trained-model at epoch 20 is much worse than the test performance by the global-trained-model at epoch 10.  But, they should be (almost) same because above code which I mentioned automatically is saving ONLY better model than previous epoch.

**Reply by holgerroth** [reply]
Can you share which job configuration you are using? For FedAvg, we do a global model selection at each round but if you use the Central job, only one round of training is performed. So the global model selection wouldn't work.

**Reply by pygabc1** [reply]
> Can you share which job configuration you are using? For FedAvg, we do a global model selection at each round but if you use the Central job, only one round of training is performed. So the global model selection wouldn't work.

Hi, Holger R. Roth, Thank you again. The configuration files are exactly same as the ones of NVFlare except only learning rate (0.01 --> 0.001), num_rounds (50 --> 2), min_clients (8 --> 2), and different model.  Do you mean that I have to use "CENTRAL=FALSE" to get best-global-model?
[https://github.com/NVIDIA/NVFlare/blob/2.2/examples/cifar10/cifar10-sim/job_configs/cifar10_fedavg/cifar10_fedavg/config/config_fed_server.json](url)
and 
[https://github.com/NVIDIA/NVFlare/blob/2.2/examples/cifar10/cifar10-sim/job_configs/cifar10_fedavg/cifar10_fedavg/config/config_fed_client.json](url) .

**Reply by holgerroth** [reply]
No, the central flag only controls which dataset split is used. For `central=True`, each client uses the full training dataset. Can you try the 2.3.0 version? I think there was an issue fixed in the global model selector which might have affected this behavior https://github.com/NVIDIA/NVFlare/pull/1401.  You should see a "best" global model in the server workspace after completion.

**Reply by pygabc1** [reply]
> No, the central flag only controls which dataset split is used. For `central=True`, each client uses the full training dataset. Can you try the 2.3.0 version? I think there was an issue fixed in the global model selector which might have affected this behavior #1401. You should see a "best" global model in the server workspace after completion.

Hi, Holger R. Roth, thank you for your reply again. As you recommend to get global model selection, I trained with “central=False”, and the following is the performance of the best_FL_global_model.pt by tested by test data. 

epochMax     AUROC
      1,               0.886
      2,               0.891
      3,              0.886
     5,               0.881
     10,             0.873

It means that the following line in def train(self,…) does not work, does it (I know the following line is updating self.best_local_model_file)?

> if val_auc_mean > self.best_acc:
>     self.best_acc = val_auc_mean
>     self.save_model(is_best=True)

Anyway, how to settle this issue (i.e., epochMax is another parameter I have to tune)? Is still using version 2.3.0 a solution (I am using 2.2, linux OS and 3.8 python)?

**Reply by holgerroth** [reply]
Thanks. Please try 2.3.0 and see if you observe the same issue.

**Reply by pygabc1** [reply]
> Thanks. Please try 2.3.0 and see if you observe the same issue.

Hi, Holger R. Roth, thank you for your reply again. I tried to use NVFlare version 2.3.0, but the result is exactly same as above, which still shows that "epochMax is another parameter I have to tune". This is from best_FL_global_model.pt.

```
epochMax AUROC
1, 0.886
2, 0.891
3, 0.886
5, 0.881
10, 0.873
```

**Reply by YuanTingHsieh** [reply]
@pygabc1

I think you mix local models with the "FL global" model.

For the logic https://github.com/NVIDIA/NVFlare/blob/2.2/examples/cifar10/pt/learners/cifar10_learner.py#L235-L239
It is for saving the local best model epoch by epoch.

For the logic https://github.com/NVIDIA/NVFlare/blob/2.2/examples/cifar10/pt/learners/cifar10_learner.py#L304-L314
It is for saving the local best model AFTER each client finish a round of training.

The above logic are all for just local model.



For "FL global" model, it is saved based on the metrics that each client return.
These lines calculate the global model evaluate on local data: https://github.com/NVIDIA/NVFlare/blob/2.2/examples/cifar10/pt/learners/cifar10_learner.py#L414-L419

This global best model is saved round by round.
If this round is better than last round, then we update it.

Your "trainer" if specified 20 epochs, because you DID NOT implement early stopping, it will finish its 20 epochs.
And the server side will get the model after 20 epochs, which has worse performance than 10 epochs.

**So in your case, your "trainer" should return the model at epoch 10 to the server.**

As for how to do "early stopping", that is the scope of Deep Learning, you can find various resources online.
